File: mobile/modules/bluetooth-sdk/android/src/main/java/com/mentra/bluetoothsdk/types/DeviceModels.kt
package com.mentra.bluetoothsdk

import com.mentra.bluetoothsdk.utils.ControllerTypes
import com.mentra.bluetoothsdk.utils.DeviceTypes

data class MentraBluetoothSdkConfig @JvmOverloads constructor(
    val deliverCallbacksOnMainThread: Boolean = true,
    val analytics: BluetoothSdkAnalyticsConfig = BluetoothSdkAnalyticsConfig(),
)

class BluetoothSdkException(
    val code: String,
    message: String,
    cause: Throwable? = null,
) : IllegalStateException(message, cause)

enum class DeviceModel(val deviceType: String) {
    G1(DeviceTypes.G1),
    G2(DeviceTypes.G2),
    MENTRA_LIVE(DeviceTypes.LIVE),
    MENTRA_NEX(DeviceTypes.NEX),
    MACH1(DeviceTypes.MACH1),
    Z100(DeviceTypes.Z100),
    FRAME(DeviceTypes.FRAME),
    NIMO(DeviceTypes.NIMO),
    AR99(DeviceTypes.AR99),
    SIMULATED(DeviceTypes.SIMULATED),
    R1(ControllerTypes.R1);

    companion object {
        @JvmStatic
        fun fromDeviceType(deviceType: String?): DeviceModel =
            values().firstOrNull { it.deviceType == deviceType } ?: MENTRA_LIVE
    }
}

data class Device(
    val model: DeviceModel,
    val name: String,
    /** Android Bluetooth address when available. */
    val address: String? = null,
    val projectName: String? = null,
    val rssi: Int? = null,
    /** Stable app-facing scan-result key. Do not parse; use typed fields instead. */
    val id: String = address?.takeIf { it.isNotBlank() } ?: "${model.deviceType}:$name",
) {
    internal fun toMap(): Map<String, Any> =
        buildMap {
            put("id", id)
            put("model", model.deviceType)
            put("name", name)
            address?.takeIf { it.isNotBlank() }?.let {
                put("address", it)
            }
            projectName?.takeIf { it.isNotBlank() }?.let {
                put("projectName", it)
            }
            rssi?.let { put("rssi", it) }
        }

    companion object {
        internal fun fromMap(values: Map<String, Any>): Device? {
            val model = stringValue(values, "model") ?: return null
            val name = stringValue(values, "name") ?: return null
            val address = stringValue(values, "address")?.takeIf { it.isNotBlank() }
            val projectName = stringValue(values, "projectName")?.takeIf { it.isNotBlank() }
            val rssi = numberValue(values, "rssi")
            val id = stringValue(values, "id")?.takeIf { it.isNotBlank() } ?: address ?: "${model}:$name"
            return Device(
                model = DeviceModel.fromDeviceType(model),
                name = name,
                address = address,
                projectName = projectName,
                rssi = rssi,
                id = id,
            )
        }
    }
}

data class ConnectOptions(
    val saveAsDefault: Boolean = true,
    val cancelExistingConnectionAttempt: Boolean = true,
)

File: mobile/modules/bluetooth-sdk/ios/Source/types/DeviceModels.swift
import Foundation

public struct MentraBluetoothSDKConfiguration {
    public static let `default` = MentraBluetoothSDKConfiguration()

    public let analytics: BluetoothSdkAnalyticsConfiguration

    public init(analytics: BluetoothSdkAnalyticsConfiguration = BluetoothSdkAnalyticsConfiguration()) {
        self.analytics = analytics
    }
}

public enum DeviceModel: String {
    case g1
    case g2
    case mentraLive
    case mentraNex
    case mach1
    case z100
    case frame
    case simulated
    case r1

    public var deviceType: String {
        switch self {
        case .g1:
            DeviceTypes.G1
        case .g2:
            DeviceTypes.G2
        case .mentraLive:
            DeviceTypes.LIVE
        case .mentraNex:
            DeviceTypes.NEX
        case .mach1:
            DeviceTypes.MACH1
        case .z100:
            DeviceTypes.Z100
        case .frame:
            DeviceTypes.FRAME
        case .simulated:
            DeviceTypes.SIMULATED
        case .r1:
            ControllerTypes.R1
        }
    }

    public static func fromDeviceType(_ deviceType: String?) -> DeviceModel {
        switch deviceType {
        case DeviceTypes.G1:
            .g1
        case DeviceTypes.G2:
            .g2
        case DeviceTypes.LIVE:
            .mentraLive
        case DeviceTypes.NEX:
            .mentraNex
        case DeviceTypes.MACH1:
            .mach1
        case DeviceTypes.Z100:
            .z100
        case DeviceTypes.FRAME:
            .frame
        case DeviceTypes.SIMULATED:
            .simulated
        case ControllerTypes.R1:
            .r1
        default:
            .mentraLive
        }
    }
}

public struct Device: Identifiable, Equatable, CustomStringConvertible {
    public let model: DeviceModel
    public let name: String
    /// CoreBluetooth identifier when available.
    public let identifier: String?
    public let rssi: Int?
    /// Stable app-facing scan-result key. Do not parse; use typed fields instead.
    public let id: String

    public init(
        model: DeviceModel,
        name: String,
        identifier: String? = nil,
        rssi: Int? = nil,
        id: String? = nil
    ) {
        self.model = model
        self.name = name
        self.identifier = identifier
        self.rssi = rssi
        self.id = id ?? identifier.flatMap { $0.isEmpty ? nil : $0 } ?? "\(model.deviceType):\(name)"
    }

    public var description: String {
        "Device(model: \(model), name: \(name))"
    }

    var dictionary: [String: Any] {
        var values: [String: Any] = [
            "id": id,
            "model": model.deviceType,
            "name": name,
        ]
        if let identifier, !identifier.isEmpty {
            values["address"] = identifier
        }
        if let rssi {
            values["rssi"] = rssi
        }
        return values
    }

    init?(values: [String: Any]) {
        guard let model = stringValue(values, "model") else { return nil }
        guard let name = stringValue(values, "name") else { return nil }
        let identifier = stringValue(values, "address").flatMap { $0.isEmpty ? nil : $0 }
        let rssi = intValue(values["rssi"])
        self.init(
            model: DeviceModel.fromDeviceType(model),
            name: name,
            identifier: identifier,
            rssi: rssi,
            id: stringValue(values, "id")
        )
    }
}

public struct ConnectOptions {
    public let saveAsDefault: Bool
    public let cancelExistingConnectionAttempt: Bool

    public init(saveAsDefault: Bool = true, cancelExistingConnectionAttempt: Bool = true) {
        self.saveAsDefault = saveAsDefault
        self.cancelExistingConnectionAttempt = cancelExistingConnectionAttempt
    }
}
